Feature: Support workload identity federation

jaredfholgate opened the pull request:
This PR is to add support for Workload Identity Federation (OIDC) in the Terraform Task.

The intention is that this will be a non-breaking change, it will simply identify the authentication scheme and setup the provider accordingly.

## Testing Evidence:

In order to run this test, the following steps were followed:

1. Compiled and packaged the extension as per the steps [HERE](https://github.com/microsoft/azure-pipelines-terraform/pull/164/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055).
2. Removed the public extension from the organisation.
3. Published the extension to the market place, shared it to the organisation and installed it in the organisation.
4. Deployed the Service Connection, Managed Identity, etc using this example code: https://github.com/Azure-Samples/azure-devops-terraform-oidc-ci-cd
5. Updated the oidc.yml pipeline to include the Terraform task steps (see below).
6. Ran the pipeline.

### Pipeline Code

![image](https://github.com/microsoft/azure-pipelines-terraform/assets/1612200/17f2531a-8da3-429d-92c3-e56e862831de)

### Test Extension

![image](https://github.com/microsoft/azure-pipelines-terraform/assets/1612200/0beabcb5-2d09-4945-9601-b8ec14c46c77)

### Service Connection

![image](https://github.com/microsoft/azure-pipelines-terraform/assets/1612200/ab7418aa-70b2-4003-b943-a4dc0dfdca6d)

### Terraform Init

![image](https://github.com/microsoft/azure-pipelines-terraform/assets/1612200/8fa97e0a-14e0-4855-be97-4360db601ae5)

### Terraform Apply

![image](https://github.com/microsoft/azure-pipelines-terraform/assets/1612200/6036e7d4-6097-48e4-85d7-7155c67f17e3)

jaredfholgate commented on Tasks/TerraformTask/TerraformTaskV4/src/azure-terraform-command-handler.ts:
```diff
@@ -39,18 +54,95 @@ export class TerraformCommandHandlerAzureRM extends BaseTerraformCommandHandler
+        workloadIdentityFederationCredentials.servicePrincipalId = tasks.getEndpointAuthorizationParameter(connectionName, "serviceprincipalid", true);
+        var connectedService: string = tasks.getInput(connectionName, true);
+        var idToken : string;
+        this.getIdToken(connectedService).then((token) => { idToken = token; });
```
My JS / TypeScript is a bit out of date. How would you call an async method from a sync method in TypeScript?

starkmsu replied:
@jaredfholgate , you can use .then( => ) method to process async calls from sync entrypoint. This is an example of such a case - https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/DockerV1/container.ts#L60

jaredfholgate commented:
@mericstam Just as a heads up on what has been done in this PR:

- Added the new functionality to enable ManagedServiceIdentity and WorkloadIdentityFederation authentication schemes to the V4 task.
- Refactored the code slightly to support async / await for the command.
- Got the unit tests working.
- Got debugging working with the unit tests.
- Fixed the existing unit tests that appeared to be broken.
- Added new unit tests for the new functionality.

In terms of what needs to happen next:

- I need to be able to deploy the task and test it end to end, as I am unable to test the token generation outside of a pipeline.
- I could get the unit tests working in the pipeline if desired, but would need access to the Azure DevOps instance.

starkmsu commented on Tasks/TerraformTask/TerraformTaskV4/src/id-token-generator.ts:
```diff
@@ -0,0 +1,49 @@
+    generate(connectedService : string): Promise<string>;
+}
+
+export class TokenGenerator implements ITokenGenerator {
```
Please use this library instead - https://github.com/microsoft/azure-pipelines-tasks-common-packages/blob/main/common-npm-packages/artifacts-common/webapi.ts#L34

jaredfholgate replied:
This is done. I have also updated the other PR to use the same method here: https://github.com/jason-johnson/azure-pipelines-tasks-terraform/pull/354/files#diff-e06f07bc6edfab65173abad59fd62d43f1fb878bc798c52e03325f7a78271a15

jaredfholgate commented:
Closing this PR in favour of https://github.com/microsoft/azure-pipelines-terraform/pull/176 which uses a local branch instead of a Fork.

jaredfholgate commented:
Closing this PR in favour of https://github.com/microsoft/azure-pipelines-terraform/pull/176 which uses a local branch instead of a Fork.
